From e944fbd24289280f774ded6cd4b13e0c0f7e1172 Mon Sep 17 00:00:00 2001
From: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
Date: Tue, 21 Nov 2023 18:06:09 +0530
Subject: [PATCH] wifi: ath12k: ignore ce interrupt until ce pipe init done

CE interrupt got trigger immediately after request irq before
disabling the interrupt, since the srng not get initialized
getting crash in ISR.

Ignoring the interrupt until the ce pipe init done.

Signed-off-by: Saravanakumar Duraisamy <quic_saradura@quicinc.com>
---
 drivers/net/wireless/ath/ath12k/ce.c   | 4 ++++
 drivers/net/wireless/ath/ath12k/core.h | 2 ++
 drivers/net/wireless/ath/ath12k/pcic.c | 4 ++++
 3 files changed, 10 insertions(+)

diff --git a/drivers/net/wireless/ath/ath12k/ce.c b/drivers/net/wireless/ath/ath12k/ce.c
index 6190730..829555d 100644
--- a/drivers/net/wireless/ath/ath12k/ce.c
+++ b/drivers/net/wireless/ath/ath12k/ce.c
@@ -1032,6 +1032,8 @@ int ath12k_ce_init_pipes(struct ath12k_base *ab)
 		}
 	}
 
+	ab->ce_pipe_init_done = true;
+
 	return 0;
 }
 
@@ -1078,6 +1080,8 @@ void ath12k_ce_free_pipes(struct ath12k_base *ab)
 			pipe->status_ring = NULL;
 		}
 	}
+
+	ab->ce_pipe_init_done = false;
 }
 
 int ath12k_ce_alloc_pipes(struct ath12k_base *ab)
diff --git a/drivers/net/wireless/ath/ath12k/core.h b/drivers/net/wireless/ath/ath12k/core.h
index 6e4a23b..392e169 100644
--- a/drivers/net/wireless/ath/ath12k/core.h
+++ b/drivers/net/wireless/ath/ath12k/core.h
@@ -1570,6 +1570,8 @@ struct ath12k_base {
 	bool multi_pd_arch;
 	/* Mode1 specific flag to identify recovering pdev */
 	bool recovery_start;
+
+	bool ce_pipe_init_done;
 	/* must be last */
 	u8 drv_priv[] __aligned(sizeof(void *));
 };
diff --git a/drivers/net/wireless/ath/ath12k/pcic.c b/drivers/net/wireless/ath/ath12k/pcic.c
index e266e8d..f47b0ee 100644
--- a/drivers/net/wireless/ath/ath12k/pcic.c
+++ b/drivers/net/wireless/ath/ath12k/pcic.c
@@ -246,6 +246,10 @@ static void ath12k_pcic_ce_tasklet(struct tasklet_struct *t)
 static irqreturn_t ath12k_pcic_ce_interrupt_handler(int irq, void *arg)
 {
 	struct ath12k_ce_pipe *ce_pipe = arg;
+	struct ath12k_base *ab = ce_pipe->ab;
+
+	if (unlikely(!ab->ce_pipe_init_done))
+		return IRQ_HANDLED;
 
 	/* last interrupt received for this CE */
 	ce_pipe->timestamp = jiffies;
-- 
2.17.1

